### ISSUE TITLE
Infinite loop with yarn workspaces + lerna + postinstall script

### ISSUE THREAD

**Issue by jsamr** [lerna/lerna#1125]
`lerna bootstrap` freeze when launched in a terminal, with yarn and workspaces enabled + a postinstall script invoking `lerna bootstrap`.

## Expected Behavior
`lerna bootstrap` terminating gracefully

## Current Behavior
`lerna bootstrap` hangs in an infinite loop

## Possible Solution

If it is a well-known bug, please warn the user if a postinstall scripts contains `lerna`.

## Steps to Reproduce (for bugs)
[Minimal working example](https://github.com/sveinburne/lerna-yarn-workspace-infinite-loop)

<details><summary>lerna.json</summary><p>

```json
{
  "lerna": "2.5.1",
  "npmClient": "yarn",
  "useWorkspaces": true,
  "version": "independent"
}
```
</p></details>

<details><summary>package.json</summary><p>

```json
{
  "name": "lerna-yarn-workspace-infinite-loop",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "postinstall": "lerna bootstrap"
  },
  "engines": {
    "node": ">=8"
  },
  "devDependencies": {
    "lerna": "^2.5.1"
  },
  "workspaces": [
    "packages/*",
    "tools/*",
    "stack"
  ]
}
```
</p></details>

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.5.1 |
| `yarn --version` | 1.3.1 |
| `node --version` | 8.9.1 |

| OS | Version |
| --- | --- |
| Ubuntu | 16.04 |

**Comment by evocateur** [comment]
Yeah, this actually is a long-standing bug when `postinstall` is used for `lerna bootstrap --hoist`, and now with `lerna bootstrap --use-workspaces`, for the same reason: both call `npm install` in the root first, which [_always_ calls `postinstall`](https://github.com/lerna/lerna/issues/1103#issuecomment-345065560).

For your use case _specifically_, using workspaces and whatnot, `yarn install` in the root is **identical** to `lerna bootstrap --use-workspaces` (indeed, lerna is [literally calling `yarn install` in the root](https://github.com/lerna/lerna/blob/48533fa3d1af3e40cd9b69dff58d4f54abd1759c/src/commands/BootstrapCommand.js#L152)). So just remove the `postinstall` and you're fine, as long as everyone (devs and CI) are using yarn (like they should).

The recursive looping from a `post*` lifecycle is annoying enough in general that I've started working on a fix that causes the `lerna` CLI to exit 0 when a lerna-provided environment variable is present, which would account for a "nested" execution during `postinstall` (while still _allowing_ `postinstall` to call `lerna` at least once).

**Comment by sveinburne** [comment]
@evocateur The fix you're working on looks great :-)
But for my specific case, I need to add `lerna run preprublish` `postinstall` script to fix #1021 ; that works gracefully.

**Comment by evocateur** [comment]
Yeah, as I mentioned in #1021, the fact that yarn doesn't run lifecycle scripts for workspaces during `yarn install` is a yarn bug, not lerna. `"postinstall": "lerna run prepublish"` would be an appropriate workaround in the meantime.
